src/mscorlib/src/System/Threading/ThreadPool.cs

```diff
@@ -458,33 +464,46 @@ internal bool LocalFindAndPop(IThreadPoolWorkItem callback)
 
         public IThreadPoolWorkItem Dequeue(ThreadPoolWorkQueueThreadLocals tl, ref bool missedSteal)
         {
-            WorkStealingQueue localWsq = tl.workStealingQueue;
             IThreadPoolWorkItem callback;
-
-            if ((callback = localWsq.LocalPop()) == null && // first try the local queue
-                !workItems.TryDequeue(out callback)) // then try the global queue
+            int wsqActiveObserved = ThreadPoolGlobals.wsqActive;
+            if(wsqActiveObserved > 0)
             {
-                // finally try to steal from another thread's local queue
-                WorkStealingQueue[] queues = WorkStealingQueueList.Queues;
-                int c = queues.Length;
-                Debug.Assert(c > 0, "There must at least be a queue for this thread.");
-                int maxIndex = c - 1;
-                int i = tl.random.Next(c);
-                while (c > 0)
+                WorkStealingQueue localWsq = tl.workStealingQueue;
+
+                if ((callback = localWsq.LocalPop()) == null && // first try the local queue
+                    !workItems.TryDequeue(out callback)) // then try the global queue
                 {
-                    i = (i < maxIndex) ? i + 1 : 0;
-                    WorkStealingQueue otherQueue = queues[i];
-                    if (otherQueue != localWsq && otherQueue.CanSteal)
+                    // finally try to steal from another thread's local queue
+                    WorkStealingQueue[] queues = WorkStealingQueueList.Queues;
+                    int c = queues.Length;
+                    Debug.Assert(c > 0, "There must at least be a queue for this thread.");
+                    int maxIndex = c - 1;
+                    int i = tl.random.Next(c);
+                    while (c > 0)
                     {
-                        callback = otherQueue.TrySteal(ref missedSteal);
-                        if (callback != null)
+                        i = (i < maxIndex) ? i + 1 : 0;
+                        WorkStealingQueue otherQueue = queues[i];
+                        if (otherQueue != localWsq && otherQueue.CanSteal)
                         {
-                            break;
+                            callback = otherQueue.TrySteal(ref missedSteal);
+                            if (callback != null)
+                            {
+                                break;
+                            }
                         }
+                        c--;
+                    }
+                    if((callback == null) && !missedSteal)
+                    {
+                        Interlocked.CompareExchange(ref ThreadPoolGlobals.wsqActive, wsqActiveObserved - 1, wsqActiveObserved);
```

Comment:
> The intended design is:

Oh right forgot about the first point. But then isn't it possible that any thread that decrements wsqActive from 1 to 0 could miss a concurrently queued work item in a local queue that it already scanned before, and it would still decrement wsqActive to 0? Then later, the thread that wakes up to satisfy the thread request would not scan to find the new work item.

> Is there a way for Enqueue() to treat as volatile w/o Dequeue() treating as volatile?

You could use Volatile.Write and a direct read on the field that is not marked as volatile.

> I am assuming there is a barrier operation included in workStealingQueue.LocalPush(). Perhaps it would be better to mark wsqActive volatile to prevent reordering of operations.

It looks like the fast path in LocalPush could just do a volatile write, that may not be sufficient for what you said.